### DISCUSSION TITLE
Error removing Cassandra for Kong Upgrade

### DISCUSSION THREAD

**Discussion by mayankmurari** [Kong/kong#13400] [Help] [answered]
We are running Kong 3.0.1 and looking to upgrade Kong post 3.4 version. We have few custom plugins where we are trying to remove cassandra present under migrations.
After removing cassandra we are getting below error while trying to restart Kong. 

`[error] init_by_lua error: /usr/local/share/lua/5.1/kong/init.lua:528: [PostgreSQL error] [PostgreSQL error] migration 'kong.plugins.jwt-auth.migrations.000_base_jwt' of 'jwt-auth' subsystem is invalid: schema violation (cassandra: required field missing)stack traceback:	[C]: in function 'assert'	/usr/local/share/lua/5.1/kong/init.lua:528: in function 'init'	init_by_lua:3: in main chunk_lua:3: in main chunk`

Below is the 000_base_jwt.lua file which we are using.
```
return {
  postgres = {
    up = [[
      CREATE TABLE IF NOT EXISTS "jwt_secrets" (
        "id"              UUID                         PRIMARY KEY,
        "created_at"      TIMESTAMP WITH TIME ZONE     DEFAULT (CURRENT_TIMESTAMP(0) AT TIME ZONE 'UTC'),
        "consumer_id"     UUID                         REFERENCES "consumers" ("id") ON DELETE CASCADE,
        "key"             TEXT                         UNIQUE,
        "secret"          TEXT,
        "algorithm"       TEXT,
        "rsa_public_key"  TEXT
      );

      DO $$
      BEGIN
        CREATE INDEX IF NOT EXISTS "jwt_secrets_consumer_id_idx" ON "jwt_secrets" ("consumer_id");
      EXCEPTION WHEN UNDEFINED_COLUMN THEN
        -- Do nothing, accept existing state
      END$$;

      DO $$
      BEGIN
        CREATE INDEX IF NOT EXISTS "jwt_secrets_secret_idx" ON "jwt_secrets" ("secret");
      EXCEPTION WHEN UNDEFINED_COLUMN THEN
        -- Do nothing, accept existing state
      END$$;
    ]],
  }
}
```

**Comment by chobits** [comment]
This error `subsystem is invalid: schema violation (cassandra: required field missing` means that you may set the `database = cassadra` string in kong.conf, while in kong 3.4, it does not support this kind of db

```
#database = postgres             # Determines the database (or no database) for
                                 # this node
                                 # Accepted values are `postgres` and `off`.
```

**Reply by chobits** [reply]
But I guess although you figure out this error problem, you still could not complete the upgrading progress directly. It needs you to migration your db from Cassadra to postgres, like extracting original configurations from original version and loading them into new kong 3.4

**Reply by mayankmurari** [reply]
@chobits We use postgres as Database and have never used cassendra. Also we have checked and there is no reference to  cassendra anywhere in the config.

**Reply by chobits** [reply]
Is jwt-auth plugin a custom plugin? I also could not find any special `cassendra` info from kong code base. So its hard for us to track why there is an error report about cassendra on 3.4 postgres.

From the following cassandra info from kong code base, kong only check it when parsing `database` option of kong.conf
```
$ grep cassandra -R kong
kong/conf_loader/constants.lua:  database = { enum = { "postgres", "cassandra", "off" }  },
kong/conf_loader/parse.lua:  if conf.database == "cassandra" then
```

**Reply by mayankmurari** [reply]
@chobits Thanks for your feedback. Tried removing cassandra with Kong upgrade version to 3.4.2 and it works that way.
So only question as cassandra was already deprecated and removed in 3.4 release so it is not possible to remove it in versions prior to 3.4?

**Reply by chobits** [reply, marked as answer]
From a technical prospective, it is viable to upgrade from Cassandra to Postgres, but debugging the upgrade problems is still challenging, so it might be better to upgrade directly to 3.4. Since we haven't maintained version 3.0.x, it's challenging for our developers to help you resolve upgrade issues unless the solution is obvious. And you could see that we are no longer maintaining the [db migration doc for 3.0.x](https://docs.konghq.com/gateway/3.0.x/migrate-cassandra-to-postgres/).
